src/pipe.js

```diff
@@ -24,15 +25,56 @@
         }
     }
 
+    function chunksCount(range) {
+        return (range[1] - range[0] + 1);
+    }
+
+    function initStateSet(state, range) {
+        var result = 0;
+        var rangeEnd = range[1];
+        for (var i = chunksCount(range) - 1; i >= 0; i--) {
+            if (initState[rangeEnd - i] === state) {
+                result++;
+            }
+        }
+        return result;
+    }
+
+    function chunkHook(attributes, hook) {
+        var range = attributes.range;
+        switch (hook) {
+            case 'onStart':
+                if (initStateSet(undefined, range) === chunksCount(range)) {
+                    ranges.push(range);
+                    hooks.onStart(attributes);
+                }
+                break;
+            case 'onBeforeInit':
+                !initStateSet('init', range) && !initStateSet('done', range) && hooks.onBeforeInit(attributes);
+                break;
+            case 'onAfterInit':
+                !initStateSet('done', range) && hooks.onAfterInit(attributes);
+                break;
+            case 'onDone':
+                var countDone = 0;
+                for (var i = ranges.length - 1; i >= 0; i--) {
+                    (initStateSet('done', ranges[i]) === chunksCount(ranges[i])) && countDone++;
+                }
+                (countDone === ranges.length) && hooks.onDone();
+                break;
+
+        }
+    }
+
     function placeholder(index) {
         placeholders[index] = currentScript();
     }
 
     function start(index, script, attributes) {
         starts[index] = currentScript();
         if (script) {
-            initState.push(index);
-            hooks.onStart(attributes);
+            chunkHook(attributes, 'onStart');
+            initState[index] = 'start';
```

Comment:
If fragment sends less link headers than the maxAssets link, this will create sparse array and its not good.